compiler/lib/pretty_print.ml

```diff
@@ -221,19 +223,35 @@ let newline st =
   st.cur <- 0; st.l <- []; st.n <- 0; st.w <- 0
 
 let to_out_channel ch =
+  let buffer = Buffer.create 100 in
   { indent = 0; box_indent = 0; prev_indents = [];
     limit = 78; cur = 0; l = []; n = 0; w = 0;
     col = 0; line = 0; total = 0;
     compact = false; pending_space = None; last_char = None; needed_space = None;
-    output = output_substring ch
+    output = (fun s i l -> Buffer.add_substring buffer s i l);
+    finally = (fun s -> 
+      Buffer.clear buffer;
+      output_string ch s);
+    get_accumulated_output = fun () -> Buffer.contents buffer;
   }
 
 let to_buffer b =
+  let buffer = Buffer.create 100 in
   { indent = 0; box_indent = 0; prev_indents = [];
     limit = 78; cur = 0; l = []; n = 0; w = 0;
     col = 0; line = 0; total = 0;
     compact = false; pending_space = None; last_char = None; needed_space = None;
-    output = fun s i l -> Buffer.add_substring b s i l }
+    output = (fun s i l -> Buffer.add_substring buffer s i l);
+    finally = (fun s -> 
+      Buffer.clear buffer;
+      Buffer.add_string b s);
+    get_accumulated_output = fun () -> Buffer.contents buffer;
```

Comment:
I just realised that the by writing on to a separate buffer the output function loses compatibility with its earlier use, as the output function is expected to write strings on to the actual out_channel/buffer. Will think it over and try a better abstraction...